composer/2022_airflow_summit/data_analytics_process_expansion.py

```diff
@@ -0,0 +1,221 @@
+        factor_sum = 0.0
+        for row in input_list:
+            latitude = row[1]
+            longitude = row[2]
+            # calculate the distance from each station to Phoenix
+            distance_to_phx = math.sqrt((phx_location[0] - latitude) ** 2 + (phx_location[1] - longitude) ** 2)
+            # calculate the distance factor of each station (1 / d^2)
+            distance_factor = 1.0 / (distance_to_phx ** 2)
+            factor_list.append(distance_factor)
+            factor_sum += distance_factor
+
+        # contains the weights of each station
+        weights_list = []
+        for val in factor_list:
+            weights_list.append(val / factor_sum)
+
+        dwa_result = 0.0
+        for row in input_list:
+            # this is the annual prcipitation/snowfall of each station
+            annual_value = row[4]
+            # weight of each station
+            weight = weights_list[input_list.index(row)]
+            dwa_result += weight * annual_value
+
+        print("weight list", weights_list)
+        print("factor list", factor_list)
+        print("return result", dwa_result)
+        print("factor sum", factor_sum)
+
+        return dwa_result
+
+    for year in range(1997, 2022):
+        # collect() function returns a list of Row object.
+        # prcp_year and snow_year will be the input for the distance weighting algorithm
+        prcp_year = (
+            annual_df.where((annual_df.ELEMENT == 'PRCP') & (annual_df.YEAR == year))
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_PRCP")).collect()
+        )
+        # print(prcp_year)
+        snow_year = (
+            annual_df.where((annual_df.ELEMENT == 'SNOW') & (annual_df.YEAR == year))
+            .groupBy("ID", "LATITUDE", "LONGITUDE", "YEAR")
+            .agg(sum("VALUE").alias("ANNUAL_SNOW")).collect()
+        )
+
+        phx_annual_prcp_df = (
+            phx_annual_prcp_df.withColumn(f"PHX_PRCP_{year}", lit(phx_dw_compute(prcp_year)))
+        )
+        phx_annual_snow_df = (
+            phx_annual_snow_df.withColumn(f"PHX_SNOW_{year}", lit(phx_dw_compute(snow_year)))
+        )
+
+    # this table has only two rows (the first row contains the columns)
+    phx_annual_prcp_df.show()
+    phx_annual_snow_df.show()
+
+    # Write results to GCS
+    if "--dry-run" in sys.argv:
+        print("Data will not be uploaded to BigQuery")
+    else:
+        # Set GCS temp location
+        temp_path = BUCKET_NAME
+
+        # Saving the data to BigQuery using the "indirect path" method and the spark-bigquery connector
+        # Uses the "overwrite" SaveMode to ensure DAG doesn't fail when being re-run
+        # See https://spark.apache.org/docs/latest/sql-data-sources-load-save-functions.html#save-modes
+        # for other save mode options
+        df.write.format("bigquery").option("temporaryGcsBucket", temp_path).mode(
+            "overwrite"
+        ).save(DF_WRITE_TABLE)
```

Comment:
Can you please reformat all of these as such:
```
(
  df.write
    .format
    .option
    .mode
    .save
)
```

Reply:
I think the best I can do is
`df.write
.format
.option
.mode
.save`

if I put all the dots in one line vertically, it will throw an `unexpected indentation` error to me.

Reply:
Done!